src/app.js

```diff
@@ -166,30 +169,70 @@ export function app() {
 			appId = message.target || '';
 			token = message.token || '';
 			debug = log('ts:io:sub:' + appId);
-			debug('CONNECTED');
+			debug('CONNECTED %o', message);
+			if (message.source) {
+				debug('SPAWNED from %o - %O', message.source, message);
+				if (methodHandlers.has('spawn')) {
+					/**
+					 * @TODO Timeout handling!
+					 */
+					methodHandlers.get('spawn').apply({}, [
+						message.data,
+						function resolve(data) {
+							postMessage({
+								type: 'SPAWN-SUCCESS',
+								target: message.source,
+								data,
+								token
+							});
+						},
+						message.source
+					]);
+				} else {
+					// this app can't be spawned and we should send an error back to the source app
+					postMessage({
+						type: 'SPAWN-FAIL',
+						target: message.source,
+						topic: message.topic,
+						data:
+							message.target +
+							" doesn't have a 'spawn' handler, it's not compatible with this SPAWN request.",
+						token
+					});
+				}
+			}
 
+			if (flushQueue(token)) {
+				debug('Publishing queued messages');
+			}
 			return;
 		}
 
 		// Call the matching handlers for the message topic.
+		if (message.type !== 'PING') {
+			debug(
+				'%s (%o) from %o - %O',
+				message.type,
+				message.topic,
+				message.source,
+				message
+			);
+		}
+
 		switch (message.type) {
 			case 'PUBLISH':
-				debug(
-					'%o (%o) from %o - %O',
-					message.type,
-					message.topic,
-					message.source,
-					message
-				);
 				handlersByTopic.forEach(
 					(handlers, topic) =>
 						matchTopic(topic, message.topic) &&
 						handlers.forEach(handler => handler(message))
 				);
 				break;
 			case 'PING':
-				postMessage({ type: 'PONG', token });
-				break;
+				return postMessage({ type: 'PONG', token });
+			case 'SPAWN-SUCCESS':
```

Comment:
same here, could be just generic 'CALL-SUCCESS' / 'CALL-FAILED'?

Reply:
SPAWN should be just a special type of CALL (has some specific parameters passed to the callee)